File: backend/src/config/databaseMode.js
/**
 * Controle de leitura/escrita para migração Firestore → MongoDB.
 *
 * DB_READ=firestore|mongo     (padrão: firestore)
 * DB_WRITE=firestore|mongo|dual (padrão: firestore)
 *
 * Exemplos:
 *   Migração paralela:  DB_READ=firestore DB_WRITE=dual
 *   Cutover leitura:    DB_READ=mongo DB_WRITE=dual
 *   Só MongoDB:         DB_READ=mongo DB_WRITE=mongo
 *   Rollback:           DB_READ=firestore DB_WRITE=firestore
 */
export function getDatabaseMode() {
  const read = (process.env.DB_READ || 'firestore').toLowerCase();
  const write = (process.env.DB_WRITE || process.env.DB_READ || 'firestore').toLowerCase();

  const readIsMongo = read === 'mongo' || read === 'mongodb';
  const writeIsMongo = write === 'mongo' || write === 'mongodb';
  const writeIsDual = write === 'dual' || write === 'both';

  const useMongo = readIsMongo || writeIsMongo || writeIsDual;
  const useFirestore = !readIsMongo || writeIsDual || write === 'firestore';

  return {
    read: readIsMongo ? 'mongo' : 'firestore',
    write: writeIsDual ? 'dual' : writeIsMongo ? 'mongo' : 'firestore',
    useMongo,
    useFirestore: useFirestore || writeIsDual || (!readIsMongo && !writeIsMongo),
  };
}

export function needsFirebaseInit() {
  return getDatabaseMode().useFirestore;
}

export function needsMongoInit() {
  return getDatabaseMode().useMongo;
}

File: backend-java/src/main/java/com/clinica/bot/config/DatabaseMode.java
package com.clinica.bot.config;

import lombok.Getter;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

@Getter
@Component
@RequiredArgsConstructor
public class DatabaseMode {

    private final ClinicaProperties properties;

    public String getRead() {
        String read = normalize(properties.getDatabase().getRead());
        return isMongo(read) ? "mongo" : "firestore";
    }

    public String getWrite() {
        String write = normalize(properties.getDatabase().getWrite());
        if ("dual".equals(write) || "both".equals(write)) {
            return "dual";
        }
        return isMongo(write) ? "mongo" : "firestore";
    }

    public boolean useMongo() {
        return "mongo".equals(getRead()) || "mongo".equals(getWrite()) || "dual".equals(getWrite());
    }

    public boolean useFirestore() {
        return !"mongo".equals(getRead()) || "dual".equals(getWrite()) || "firestore".equals(getWrite());
    }

    public boolean needsFirebaseInit() {
        return useFirestore();
    }

    public boolean needsMongoInit() {
        return useMongo();
    }

    private static String normalize(String value) {
        return value == null ? "" : value.trim().toLowerCase();
    }

    private static boolean isMongo(String value) {
        return "mongo".equals(value) || "mongodb".equals(value);
    }
}
